runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt

```diff
@@ -0,0 +1,146 @@
+        assertEquals(Json.stringify(data2), JSON.stringify(serialized2), "should serialize characters as strings")
+        assertEquals(data2.a.toString(), serialized2.a) // FIXME why is toString needed here, both should be char?
+
+        val data3 = AllTypes(1, 2, 3, 4.0f, 5.0, 'c', true, "string")
+        val serialized3 = DynamicObjectSerializer().serialize(data3)
+        assertEquals(Json.stringify(data3), JSON.stringify(serialized3))
+    }
+
+    @Test
+    fun wrappedObjectsTest() {
+        val data = DataWrapper("a string", Data(42))
+        val serialized = DynamicObjectSerializer().serialize(data)
+        assertEquals(Json.stringify(data), JSON.stringify(serialized))
+        assertEquals(data.s, serialized.s)
+        assertNotNull(serialized.d)
+        assertEquals(data.d?.a, serialized.d.a)
+    }
+
+    @Test
+    fun nullsTest() {
+        val dataNull = DataWrapper("a string", null)
+        val serializedNull = DynamicObjectSerializer().serialize(dataNull)
+        assertEquals(dataNull.s, serializedNull.s)
+        assertNull(serializedNull.d)
+        assertFalse(js("""Object.keys(serializedNull).includes("d")"""), "should omit null properties")
+    }
+
+    @Test
+    fun listTest() {
+        val data = listOf(1, 2, 3, 44)
+        val serialized = DynamicObjectSerializer().serialize(IntSerializer.list, data)
+        assertEquals(Json.stringify(IntSerializer.list, data), JSON.stringify(serialized))
+        assertEquals(data.first(), serialized[0])
+        assertEquals(data.last(), serialized[3])
+        assertNotNull(serialized.length, "lenght property should exist")
+        assertEquals(data.size, serialized.length)
+    }
+
+    @Test
+    fun arrayTest() {
+        val data = intArrayOf(1, 2, 3, 44)
+
+        val serialized = DynamicObjectSerializer().serialize(IntArraySerializer, data)
+        assertEquals(Json.stringify(IntArraySerializer, data), JSON.stringify(serialized))
+        assertEquals(1, serialized[0])
+        assertEquals(44, serialized[3])
+        assertNotNull(serialized.length, "lenght property should exist")
+    }
+
+    @Test
+    fun nestedListTest() {
+        val data = NestedList("a string", listOf(1, 2, 3, 44))
+        val serialized = DynamicObjectSerializer().serialize(data)
+        assertEquals(Json.stringify(data), JSON.stringify(serialized))
+        assertEquals(data.a, serialized.a)
+        assertNotNull(serialized.list)
+        assertEquals(data.list[0], serialized.list[0])
+        assertEquals(data.list[1], serialized.list[1])
+        assertEquals(data.list[2], serialized.list[2])
+        assertEquals(data.list[3], serialized.list[3])
+    }
+
+    @Test
+    fun complexMapWrapperTest() {
+        val data = ComplexMapWrapper(mapOf("key1" to Data(1), "key2" to Data(2)))
+        val serialized = DynamicObjectSerializer().serialize(data)
+        assertEquals(Json.stringify(data), JSON.stringify(serialized))
```

Comment:
You can reduce copy-paste by extracting these 3 lines into a function similar to `assertStringFormAndRestored` (e.g. `assertDynamicForm`)